OneDriveSDK/Auth/ODBaseAuthProvider.m

```diff
@@ -94,27 +100,34 @@ - (void)authorizationFlowCompletedWithURL:(NSURL *)endURL
                  presentingViewController:(UIViewController *)presentingViewController
                                completion:(AuthCompletion)completionHandler
 {
-    // Always remove the auth view when we finished loading.
-    dispatch_async(dispatch_get_main_queue(), ^{
-        [presentingViewController dismissViewControllerAnimated:NO completion:nil];
-    });
-    
     if (!error){
         [self.logger logWithLevel:ODLogDebug message:@"Response from auth service", endURL];
         NSString *code = [ODAuthHelper codeFromCodeFlowRedirectURL:endURL];
         if (code){
-            [self getTokenWithCode:code completion:completionHandler];
+            [self getTokenWithCode:code completion:^(NSError *error) {
+                [self dismissAndCompleteAuthorizationWithError:error completion:completionHandler];
+            }];
         }
         else{
             [self.logger logWithLevel:ODLogError message:@"Error reading code from response"];
-            completionHandler([self errorFromURL:endURL]);
+            [self dismissAndCompleteAuthorizationWithError:[self errorFromURL:endURL] completion:completionHandler];
         }
     }
     else{
-        completionHandler(error);
+        [self dismissAndCompleteAuthorizationWithError:error completion:completionHandler];
     }
 }
 
+- (void)dismissAndCompleteAuthorizationWithError:(NSError*)error completion:(AuthCompletion)completionHandler {
+    __block ODAuthenticationViewController *authVC = self.authViewController;
```

Comment:
__block shouldn't be necessary here. You aren't modifying authVC in the block or expecting modifications to carry outside of the block. A copy should be fine.